src/psp/guest_request.rs

```diff
@@ -0,0 +1,531 @@
+use x86_64::{PhysAddr, VirtAddr};
+
+// AEAD Algo
+#[allow(dead_code)]
+const SNP_AEAD_INVALID: u8 = 0;
+const SNP_AEAD_AES_256_GCM: u8 = 1;
+
+// SNP_GUEST_MESSAGE type
+pub const SNP_MSG_TYPE_INVALID: u8 = 0;
+
+const HDR_VERSION: u8 = 1;
+const MSG_VERSION: u8 = 1;
+
+const AUTHTAG_SIZE: i32 = 16;
+const DEFAULT_IV_SIZE: usize = 12;
+const KEY_SIZE: usize = 32;
+const U64_SIZE: usize = core::mem::size_of::<u64>();
+const MSG_HDR_SIZE: usize = core::mem::size_of::<SnpGuestRequestMsgHdr>();
+
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone)]
+struct SnpGuestRequestMsgHdr {
+    authtag: [u8; 32usize],
+    msg_seqno: u64,
+    rsvd1: [u8; 8usize],
+    algo: u8,
+    hdr_version: u8,
+    hdr_sz: u16,
+    msg_type: u8,
+    msg_version: u8,
+    msg_sz: u16,
+    rsvd2: u32,
+    msg_vmpck: u8,
+    rsvd3: [u8; 35usize],
+}
+
+#[repr(C, packed)]
+#[derive(Debug, Copy, Clone)]
+struct SnpGuestRequestMsg {
+    hdr: SnpGuestRequestMsgHdr,
+    payload: [u8; 4000usize],
+}
+
+type SnpGuestRequestMsgPtr = *mut SnpGuestRequestMsg;
+type EvpCipherCtxPtr = *mut EVP_CIPHER_CTX;
+
+impl SnpGuestRequestMsg {
+    fn alloc_raw() -> SnpGuestRequestMsgPtr {
+        let pa: PhysFrame = match mem_allocate_frame() {
+            Some(f) => f,
+            None => return ptr::null_mut(),
+        };
+        let va: VirtAddr = pgtable_pa_to_va(pa.start_address());
+        pgtable_make_pages_shared(va, PAGE_SIZE);
+        memset(va.as_mut_ptr(), 0, PAGE_SIZE as usize);
+        va.as_mut_ptr() as SnpGuestRequestMsgPtr
+    }
+    fn free_raw(p: SnpGuestRequestMsgPtr) {
+        let va: VirtAddr = VirtAddr::from_ptr(p);
+        pgtable_make_pages_private(va, PAGE_SIZE);
+        free_page(va);
+    }
+}
+
+static SEQ_NUM: AtomicU64 = AtomicU64::new(0);
+fn seqno_last_used() -> u64 {
+    SEQ_NUM.load(Ordering::Relaxed)
+}
+fn seqno_add_two() {
+    SEQ_NUM.fetch_add(2, Ordering::Relaxed);
+}
+
+struct SnpGuestRequestCmd {
+    // The Guest Request NAE event requires two unique pages: one for
+    // the request and another for the response message. Both of them
+    // are assigned to the hypervisor (shared).
+    // We encrypt the request payload using AES-256-GCM and the vmpck[0]
+    // key before sending it. The PSP firmware does the same with the
+    // response payload.
+    req: SnpGuestRequestMsgPtr,
+    resp: SnpGuestRequestMsgPtr,
+    ctx_req: EvpCipherCtxPtr,
+    ctx_resp: EvpCipherCtxPtr,
+}
+
+// The SnpGuestRequestCmd fields are raw pointers and we need to properly
+// free them.
+impl Drop for SnpGuestRequestCmd {
+    fn drop(&mut self) {
+        self.free();
+    }
+}
+
+impl SnpGuestRequestCmd {
+    fn new() -> Self {
+        SnpGuestRequestCmd {
+            req: ptr::null_mut(),
+            resp: ptr::null_mut(),
+            ctx_req: ptr::null_mut(),
+            ctx_resp: ptr::null_mut(),
+        }
+    }
+
+    fn init(&mut self) -> bool {
+        self.req = SnpGuestRequestMsg::alloc_raw();
```

Comment:
By having the struct members be VirtAddr's you can keep the logic around allocation and making the page shared here in the init function where it's a bit more obvious.